### PR TITLE
 Introduce Errorable Pattern, Shell Integration, Add Test

### PR DESCRIPTION
**Pull request by Tatsinnit**
This PR refactors error handling and shell command execution by introducing the **Errorable pattern**  also takes care of idea proposed here #21  and implementing shell for cross-platform compatibility. Gentle fyi @cpuguy83 and @ashu8912  ❤️ 

This PR Adds type-safe error handling via the `Errorable` pattern and replaces shell command execution with secure array-based arguments to prevent injection vulnerabilities.

### Key Changes

**New utilities:**

- `errorable.ts` - Type-safe error handling with `Succeeded<T> | Failed` discriminated unions
- shell.ts - Secure command execution using `child_process.execFile` with array arguments

**Updated code:**
- targetHelpers.ts, `documentHelpers.ts`, `runBuildCommand.ts` - Now return `Errorable<T>` for type-safe errors

**Security improvements:**
- After: `execFile('docker', ['build', '-f', filePath])` - injection-proof

**Dependencies:**
- Removed `shelljs` and `@types/shelljs`
- Using built-in Node.js `child_process` only

**Testing:**

- 55 tests passing (34 errorable + 21 shell execution tests)
- Covers security, error handling, and cross-platform compatibility

### Benefits

- **Security**: Array-based arguments prevent shell injection
- **Type safety**: Explicit success/failure states with TypeScript discrimination
- **Simplicity**: Fewer dependencies, cleaner error handling
- **Composability**: Functional helpers (`map`, `bind`, `combine`) for error propagation

### Example

```typescript
const result = await execFile('docker', ['buildx', 'build', '-f', specPath], {
  cwd: workspaceRoot,
  env: { BUILDX_EXPERIMENTAL: '1' }
});

if (failed(result)) {
  vscode.window.showErrorMessage(result.error);
  return;
}

const output = result.result.stdout; // Type-safe access
```

### 1. Errorable Type Pattern

Created `src/commands/utils/errorable.ts` implementing a type-safe error handling pattern:

```typescript
export type Errorable<T> = Succeeded<T> | Failed;

export interface Succeeded<T> {
  readonly succeeded: true;
  readonly result: T;
}

export interface Failed {
  readonly succeeded: false;
  readonly error: string;
}
```

**Benefits:**
- Type-safe error handling
- Explicit success/failure states
- Better composability with functional programming patterns
- Rich helper functions (`map`, `bind`, `combine`, etc.)
- Consistent error propagation across the codebase

### 2. Secure Shell Execution

Created `src/commands/utils/shell.ts` using `child_process.execFile` for secure command execution:

**Before:**
```typescript
import { execFile } from 'child_process';
import { promisify } from 'util';

const execFileAsync = promisify(execFile);
const { stdout } = await execFileAsync('docker', args, execOptions);
```

**After:**
```typescript
import { execFile } from '../../utils/shell';
import { failed } from '../../utils/errorable';

const shellResult = await execFile('docker', args, {
  cwd: contextPath,
  env: { BUILDX_EXPERIMENTAL: '1' },
});
if (failed(shellResult)) {
  // Handle error
  return;
}
const output = shellResult.result.stdout;
```

**Benefits:**
- **Security**: Array-based arguments prevent shell injection
- Cross-platform compatibility (Windows, macOS, Linux)
- Simpler API with unified error handling
- Better integration with Errorable pattern
- No shell interpretation of special characters
- No need for promisify wrappers
- Consistent async/sync execution options

**Available Functions:**
- `execFile(file, args, options)` - Secure array-based execution (async)
- `execFileSync(file, args, options)` - Synchronous version of execFile

#### `targetHelpers.ts`
- Replaced `execFile` with `exec` from shell utilities
- Returns `Errorable<BuildTargetInfo[]>` for type-safe error handling
- Improved error messages with `getDockerErrorMessage`

#### `documentHelpers.ts`
- Updated `extractDalecSpecMetadata` to return `Errorable<DalecSpecMetadata>`
- Better error propagation with descriptive error messages

#### `runBuildCommand.ts`
- Updated to handle Errorable return types
- Added `failed()` checks for proper error handling

### 4. Dependencies

No external dependencies added - uses built-in Node.js `child_process` module.

### Using Errorable Pattern

```typescript
import { Errorable, failed, succeeded } from '../utils/errorable';

async function myFunction(): Promise<Errorable<string>> {
  try {
    const result = await someOperation();
    return { succeeded: true, result };
  } catch (error) {
    return { succeeded: false, error: getErrorMessage(error) };
  }
}

// Using the result
const result = await myFunction();
if (failed(result)) {
  console.error(result.error);
  return;
}
console.log(result.result); // Type-safe access
```

### Using Shell Utilities

**Secure array-based execution (recommended):**
```typescript
import { execFile, execFileSync, NonZeroExitCodeBehaviour } from '../utils/shell';
import { failed } from '../utils/errorable';

// Async execution with array-based arguments - no shell injection risk
const shellResult = await execFile('docker', ['buildx', 'build', '-f', filePath, contextPath], {
  cwd: '/path/to/workdir',
  env: { CUSTOM_VAR: 'value' },
  exitCodeBehaviour: NonZeroExitCodeBehaviour.Fail
});

if (failed(shellResult)) {
  console.error('Command failed:', shellResult.error);
  return;
}

console.log('Output:', shellResult.result.stdout);
console.log('Exit code:', shellResult.result.code);

// Synchronous execution
const syncResult = execFileSync('docker', ['--version']);
if (failed(syncResult)) {
  console.error('Failed to get Docker version:', syncResult.error);
}
```

## Helper Functions

The Errorable pattern includes several utility functions:

- `failed<T>(e: Errorable<T>)`: Type guard for failed results
- `succeeded<T>(e: Errorable<T>)`: Type guard for successful results
- `map<T, U>(e: Errorable<T>, fn: (t: T) => U)`: Transform successful values
- `bind<T, U>(e: Errorable<T>, fn: (t: T) => Errorable<U>)`: Chain operations
- `combine<T>(es: Errorable<T>[])`: Combine multiple results
- `getErrorMessage(error: unknown)`: Extract error messages safely

### Run tests result:

Screenshot 2026-01-01 at 7 48 07 PM

Screenshot 2026-01-01 at 7 47 53 PM

### CODE REVIEW

File: src/commands/utils/shell.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    } else {
+      return {
+        succeeded: false,
+        error: `Command "${cmd}" failed with exit code ${result.code}.\nStdout:\n${result.stdout}\nStderr:\n${result.stderr}`,
+      };
+    }
+  } catch (ex) {
+    return { succeeded: false, error: getErrorMessage(ex) };
+  }
+}
```
[the remark this answers is not in the document]
Comment (Tatsinnit) [reply src/commands/utils/shell.ts]: Instead of string concatenation, use the array-based approach. let me see.

File: src/commands/utils/shell.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        };
+      }
+    } finally {
+      shell.config.silent = prevSilent;
+      shell.config.fatal = prevFatal;
+    }
+  } catch (ex) {
+    return { succeeded: false, error: getErrorMessage(ex) };
+  }
+}
```
[the remark this answers is not in the document]
Comment (Tatsinnit) [reply src/commands/utils/shell.ts]: Added tests now. therefore resolving. Thanks

**Comment by Tatsinnit** [comment]
Thank you, so far!! =) @copilot please review this PR focusing on:

- correctness
- safety
- obvious bugs or edge cases

No style or nitpick feedback needed. Thanks!

File: src/commands/utils/errorable.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+export function failed<T>(e: Errorable<T>): e is Failed {
+    return !e.succeeded;
+}
+
+export function success<T>(item: T): Errorable<T> {
+    return { succeeded: true, result: item };
+}
+
```
Comment (ashu8912) [inline src/commands/utils/errorable.ts]: we have success do you think we should also add

```
export function failure(error: string): Failed {
    return { succeeded: false, error };
}
```

File: src/commands/runBuildCurrentSpecCommand/helpers/targetHelpers.ts [hunk 4]
```diff
@@ -111,17 +110,26 @@ export async function getFrontendTargets(document: vscode.TextDocument): Promise
       try {
```
Comment (ashu8912) [inline src/commands/runBuildCurrentSpecCommand/helpers/targetHelpers.ts]: Tats should we remove this try catch now since we use the errorable pattern to handle errors?

**Review by ashu8912 — COMMENTED** [review]
Added some questions and comments

**Review by cpuguy83 — APPROVED** [review]
Definitely not TS person but this LGTM.
